src/js/angular/core/services/jwt-auth.service.js

```diff
@@ -3,294 +3,302 @@ import 'angular/core/services/openid-auth.service.js';
+             */
+            $rootScope.setPermissionDenied = (path) => {
+                if (path === LOGIN_PATH || !isAuthenticated()) {
                     return false;
                 }
                 $rootScope.deniedPermissions[path] = true;
                 return true;
             };
-            $rootScope.hasPermission = function () {
+
+            /**
+             * Checks if permission is granted for the current path.
+             * @return {boolean} True if permission is granted.
+             */
+            $rootScope.hasPermission = () => {
                 const path = $location.path();
                 return !$rootScope.deniedPermissions[path];
             };
 
-            this.updateReturnUrl = () => {
-                if ($location.url().indexOf('/login') !== 0) {
-                    $rootScope.returnToUrl = $location.url();
-                }
-            };
-
             /**
              * Redirects to the login page.
-             *
              * @param {boolean} expired If true a toast message about expired login will be shown.
-             *                          In some cases this will be autodetected.
              * @param {boolean} noaccess If true a toast message about no access/config error will be shown.
-             *                           This overrides 'expired'.
              * @return {Promise<unknown>}
              */
-            $rootScope.redirectToLogin = function (expired, noaccess) {
-                if (jwtAuth.authTokenIsType('Bearer')) {
+            $rootScope.redirectToLogin = (expired, noaccess) => {
+                if (_authTokenIsType(TOKEN_TYPE_BEARER)) {
                     // OpenID login may be detected as expired either on initial validation
                     // when we get expired = true or indirectly via 401, setting expired = true
                     // handles the second case.
                     expired = true;
                 }
-                jwtAuth.clearAuthenticationInternal();
+                _clearAuthenticationInternal();
                 // remember where we were so we can return there
-                jwtAuth.updateReturnUrl();
+                updateReturnUrl();
 
-                $location.path('/login');
+                $location.path(LOGIN_PATH);
                 if (noaccess) {
-                    $location.search('noaccess');
+                    $location.search(NO_ACCESS);
                 } else if (expired) {
-                    $location.search('expired');
+                    $location.search(EXPIRED);
                 }
                 // Countering race condition. When the unauthorized interceptor catches error 401 or 409, then we must make
                 // sure that a request is made to access the login page before proceeding with the rejection of the
                 // original request. Otherwise the login page is not accessible in the context of spring security.
```

Comment:
Nice. A simple thing like the logout tuned into a nightmare. There is a recent bug about this BTW.